### PR TITLE
[pipelines] add regression analysis and KPIs

### PR DESCRIPTION
**Pull request by drewrip**

### CODE REVIEW

File: projects/pipelines/testing/config.yaml [hunk 1]
```diff
@@ -160,6 +160,17 @@ ci_presets:
     tests.pipelines.project_count: 5
     tests.pipelines.capture_extra_artifacts: false
 
+  gating:
```
Comment (kpouget) [inline projects/pipelines/testing/config.yaml:163]: can you rename this maybe `gating_test`?

**Review by kpouget — COMMENTED** [review]
thanks Drew! 
can you launch the OCP CI `rhoai-light` test with the `gating` preset? if anything part of our current `gating` preset is mandatory, include in in a `gating_smoke` test, so that we have an easy way to rehearse the the regression test in the CI

let's see if things go well and the regression report is generated. From there we'll review the KPIs and labels. Looking forward to it :D

File: projects/pipelines/testing/config.yaml (hunk 1 above)
Comment (drewrip) [reply projects/pipelines/testing/config.yaml]: Done!

**Comment by drewrip** [comment]
@kpouget Definitely! Let me add a `gating_smoke` preset that's more suitable for this. I have a feeling a few things are probably missing, but I think most of the coding is done

**Comment by kpouget** [comment]
hum, the test failed because of [this line](https://github.com/openshift-psap/topsail/blob/8cf516ac14bc8ff27b28983b08196c31e2753328/projects/pipelines/visualizations/rhods-pipelines/store/parsers.py#L433-L434)
```
                    all_workflow_start_times[f"{name}"] = datetime.datetime.strptime(
                        root_node["startedAt"], helpers_store_parsers.K8S_TIME_FMT)
```
because the workflow never started I guess :/

**Comment by kpouget** [comment]
looks good @drewrip , next test shoud have a regression report :)

**Comment by kpouget** [comment]
**Test of 'rhoai test test_ci' failed after 01 hours 08 minutes 10 seconds.**

* Link to the [test results](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836480196228681728/artifacts/e2e/test/artifacts/).
* Link to the [reports index](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836480196228681728/artifacts/e2e/test/artifacts//reports_index.html).


**Test configuration**:
```
# RHOAI: run pipelines test test_ci
PR_POSITIONAL_ARGS=gating_smoke
PR_POSITIONAL_ARG_0=pipelines-rhoai-e2e
PR_POSITIONAL_ARG_1=gating_smoke
```

*[TOPSAIL auto-generated message]*

**Comment by drewrip** [comment]
> hum, the test failed because of [this line](https://github.com/openshift-psap/topsail/blob/8cf516ac14bc8ff27b28983b08196c31e2753328/projects/pipelines/visualizations/rhods-pipelines/store/parsers.py#L433-L434)
> 
> ```
>                     all_workflow_start_times[f"{name}"] = datetime.datetime.strptime(
>                         root_node["startedAt"], helpers_store_parsers.K8S_TIME_FMT)
> ```
> 
> because the workflow never started I guess :/

looks like the runs in OpenShift CI are failing because of this, but tests in Jenkins don't show this issue. The pods that make up the Workflows (the actual pipeline pods) are stuck pending. I need to see why

**Comment by kpouget** [comment]
for the `Pending` status, you need to look at the Pod `describe`:
```
  Warning  FailedScheduling  26m (x19 over 59m)  default-scheduler  0/7 nodes are available: 1 Insufficient cpu, 1 node(s) had untolerated taint {only-test-pods: yes}, 2 node(s) didn't match Pod's node affinity/selector, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/7 nodes are available: 1 No preemption victims found for incoming pod, 6 Preemption is not helpful for scheduling.
```
--> there's no CPU available

I guess that's because there's only 1 `m6i.2xlarge` [machine](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836480196228681728/artifacts/e2e/prepare/artifacts/000__prepare_cluster/001__cluster__set_scale/machines_rhods-pipelines.status)
```
NAME                    PHASE     TYPE          REGION      ZONE         AGE
rhods-pipelines-d4czv   Running   m6i.2xlarge   us-west-2   us-west-2a   5m9s
```
I'm not sure anymore how this `1 machine` is configured. In the notebook scale test, I had a computation that was taking into account the number of users and the requested size per notebook, and the cpu/memory of the instance type.
That was quite reliable to scale up the cluster

**Comment by drewrip** [comment]
> for the `Pending` status, you need to look at the Pod `describe`:
> 
> ```
>   Warning  FailedScheduling  26m (x19 over 59m)  default-scheduler  0/7 nodes are available: 1 Insufficient cpu, 1 node(s) had untolerated taint {only-test-pods: yes}, 2 node(s) didn't match Pod's node affinity/selector, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }. preemption: 0/7 nodes are available: 1 No preemption victims found for incoming pod, 6 Preemption is not helpful for scheduling.
> ```
> 
> --> there's no CPU available
> 
> I guess that's because there's only 1 `m6i.2xlarge` [machine](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836480196228681728/artifacts/e2e/prepare/artifacts/000__prepare_cluster/001__cluster__set_scale/machines_rhods-pipelines.status)
> 
> ```
> NAME                    PHASE     TYPE          REGION      ZONE         AGE
> rhods-pipelines-d4czv   Running   m6i.2xlarge   us-west-2   us-west-2a   5m9s
> ```
> 
> I'm not sure anymore how this `1 machine` is configured. In the notebook scale test, I had a computation that was taking into account the number of users and the requested size per notebook, and the cpu/memory of the instance type. That was quite reliable to scale up the cluster

Yeah this hasn't been something I've tuned yet, so whatever the previous instance requirements were are still there. We ran into this a bit during the DSP tests during the last release.  What I've just realized is that the per user resource requirements are likely much larger for DSP than notebooks. It is likely also more difficult to estimate CPU and memory requirements just based on `user_count`. Each user starts a pipeline that attempts to spawn an arbitrary number of pods as part of the workflow. So the number of pods, and the corresponding resources needed to accommodate them, depends on both the `user_count` and each specific pipeline topology.

**Comment by kpouget** [comment]
**Test of 'rhoai test test_ci' succeeded after 00 hours 17 minutes 57 seconds.**

* Link to the [test results](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836805056121278464/artifacts/e2e/test/artifacts/).
* Link to the [reports index](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836805056121278464/artifacts/e2e/test/artifacts//reports_index.html).


**Test configuration**:
```
# RHOAI: run pipelines test test_ci
PR_POSITIONAL_ARGS=gating_smoke
PR_POSITIONAL_ARG_0=pipelines-rhoai-e2e
PR_POSITIONAL_ARG_1=gating_smoke
```

*[TOPSAIL auto-generated message]*

**Comment by drewrip** [comment]
@kpouget Looks like increasing the resource requirements associated with each user (for the SUT nodes) fixed the problem. We are getting [a nice regression report](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836805056121278464/artifacts/e2e/test/artifacts/005__plots/regression_report.html) now :). Good to merge from my end

**Comment by kpouget** [comment]
thanks @drewrip , looks great!

a few minor details:

in the KPI labels, 
* `run_count` comes from `pipelines_per_use`. Why not using  `pipelines_per_use` as in the config?
* `runs_per_pipeline` is configured, but not exposed as a label. Do you think it won't ever change in CPT?

and, very minor:
```
results.sleep_factor = int(results.test_config.get("tests.pipelines.sleep_factor"))
```
when you read from the configuration, the YAML parser already casted to the right type, so no need to cast again to `int`
(unless you hit an issue around that? 🤔 )

**Comment by drewrip** [comment]
> in the KPI labels,
> 
>     run_count comes from pipelines_per_use. Why not using pipelines_per_use as in the config?
>     runs_per_pipeline is configured, but not exposed as a label. Do you think it won't ever change in CPT?

That's a good question. Originally I wanted to rename "runs_per_pipeline" to "run_count" in the labels because I thought that was more concise (and outwardly consumable) but I'm just going to keep "runs_per_pipeline" as I don't think its worth the confusion.

> and, very minor:
> 
> results.sleep_factor = int(results.test_config.get("tests.pipelines.sleep_factor"))
> 
> when you read from the configuration, the YAML parser already casted to the right type, so no need to cast again to int
> (unless you hit an issue around that? 🤔 )

This was just a personal preference, but I'm happy to remove the casts to attempt to follow the style of the rest of the repo. Having the cast makes it extra clear to myself and my type checker what types should be expected since we annotate types everywhere.

**Comment by kpouget** [comment]
**Test of 'rhoai test test_ci' succeeded after 00 hours 14 minutes 43 seconds.**

* Link to the [test results](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836871773979676672/artifacts/e2e/test/artifacts/).
* Link to the [reports index](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836871773979676672/artifacts/e2e/test/artifacts//reports_index.html).


**Test configuration**:
```
# RHOAI: run pipelines test test_ci
PR_POSITIONAL_ARGS=gating_smoke
PR_POSITIONAL_ARG_0=pipelines-rhoai-e2e
PR_POSITIONAL_ARG_1=gating_smoke
```

*[TOPSAIL auto-generated message]*

**Comment by drewrip** [comment]
[The labels are more complete now](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/499/pull-ci-openshift-psap-topsail-main-rhoai-e2e/1836871773979676672/artifacts/e2e/test/artifacts/005__plots/regression_report.html). Let me know what you think. Good to merge from my end

**Comment by kpouget** [comment]
thanks @drewrip , good to merge!
/lgtm
I let you see if you want to squash or keep the commits

**Comment by drewrip** [comment]
@kpouget squashed a few of the commits and rebased, could you give another lgtm?

**Comment by kpouget** [comment]
thanks for the work @drewrip , looking forward to seeing true regression reports :)
/lgtm

**Pull request merged**